server/src/main/java/org/apache/druid/server/SubqueryLimitUtils.java

```diff
@@ -0,0 +1,132 @@
+  )
+  {
+    autoLimitBytes = computeLimitBytesForAuto(lookupManager, maxMemoryInJvm, brokerNumHttpConnections);
+  }
+
+  public long convertSubqueryLimitStringToLong(final String maxSubqueryLimit)
+  {
+    if (UNLIMITED_LIMIT_VALUE.equalsIgnoreCase(maxSubqueryLimit)) {
+      return UNLIMITED_LIMIT_REPRESENTATION;
+    }
+    if (AUTO_LIMIT_VALUE.equalsIgnoreCase(maxSubqueryLimit)) {
+      return autoLimitBytes;
+    }
+
+    long retVal;
+    try {
+      retVal = Long.parseLong(maxSubqueryLimit);
+    }
+    catch (NumberFormatException e) {
+      throw InvalidInput.exception(
+          e,
+          "Unable to parse the provided maxSubqueryLimit [%s] to a valid number. Valid values for the "
+          + "maxSubqueryLimits can be 'auto', 'unlimited' or a positive number representing bytes to reserve.",
+          maxSubqueryLimit
+      );
+    }
+
+    // This can happen if the provided number is greater than Longs.MAX_VALUE
+    if (retVal < 0) {
+      throw InvalidInput.exception("Limit too large");
+    }
+
+    return retVal;
+  }
+
+  /**
+   * Computes the byte limit when 'auto' is passed as a parameter. This computes the total heap space available
+   * for the subquery inlining by getting a fraction of the total heap space in JVM, removing the size of the lookups,
+   * and dividing it by the maximum concurrent queries that can run. Maximum concurrent queries that Druid can
+   * run is usually limited by its broker's http threadpool size
+   */
+  private static long computeLimitBytesForAuto(
+      final LookupExtractorFactoryContainerProvider lookupManager,
+      final long maxMemoryInJvm,
+      final int brokerNumHttpConnections
+  )
+  {
+    long memoryInJvmWithoutLookups = maxMemoryInJvm - computeLookupFootprint(lookupManager);
+    long memoryInJvmForSubqueryResultsInlining = (long) (memoryInJvmWithoutLookups * SUBQUERY_MEMORY_BYTES_FRACTION);
+    long memoryInJvmForSubqueryResultsInliningPerQuery = memoryInJvmForSubqueryResultsInlining
+                                                         / brokerNumHttpConnections;
+    return Math.max(memoryInJvmForSubqueryResultsInliningPerQuery, 1L);
+  }
+
+  /**
+   * Computes the size occupied by the lookups. If the size of the lookup cannot be computed, it skips over the lookup
+   */
+  private static long computeLookupFootprint(final LookupExtractorFactoryContainerProvider lookupManager)
+  {
+
+    if (lookupManager == null || lookupManager.getAllLookupNames() == null) {
+      log.warn("Failed to get the lookupManager for estimating lookup size. Skipping.");
+      return 0;
+    }
+
+    int lookupCount = 0;
+    long lookupFootprint = 0;
+
+    for (final String lookupName : lookupManager.getAllLookupNames()) {
+      final LookupExtractorFactoryContainer container = lookupManager.get(lookupName).orElse(null);
+
+      if (container != null) {
+        try {
+          final LookupExtractor extractor = container.getLookupExtractorFactory().get();
```

Comment:
Questions:
What ensures lookups are loaded completely.
`lookupFootprint += extractor.estimateHeapFootprint();` seems to be traversing each key value pair so it can become slow leading to slower broker startups times no ?

Reply:
1. Lookups are usually loaded completely on the startup for querying processes. The behavior is handled by the flag - `enableLookupSyncOnStartup` which ensures that it is ideally set to true for the brokers.
2. Considering that a single entry takes about 100 bytes in the map table, a 100KB lookup would have about 1000 entries. Iterating over it a single time should be fine I think. For larger lookups, however, say unusual cases where the size of the lookup is over a limit, we can have some heuristics instead.
3. Since we do `estimateHeapFootprint` for every MSQ query, I think it should be relatively safe to include one time at the broker startup time. WDYT?

Reply:
Since MSQ is not time sensitive currently and the broker start up is we might need better logging here. 
So here is what I propose. 
Add an info log line mentioning the start time for this calculation and the end time for this calculation. Also mention in the log line that if this delta time is > 10 seconds, consider lowering the size of the lookups. 
If there is a way to cache the size information on the lookup side that will be even cooler but I guess that can be a follow up PR to this one. 
cc @pranavbhole who is working in the lookups area.